File: src/main/kotlin/io/github/sumaroder/rosenfeld/tokenize/Tokenizer.kt
package io.github.sumaroder.rosenfeld.tokenize

import io.github.sumaroder.rosenfeld.error.ErrorHandler

class CharStream(val src: String) {
    private val length: Int = src.length
    private var idx: Int = 0
    var line: Int = 1
    var column: Int = 1
    private val linesStartIndices: List<Int>

    init {
        val indices = mutableListOf(0)
        for (i in src.indices) {
            if (src[i] == '\n') {
                indices.add(i + 1)
            }
        }
        linesStartIndices = indices
    }

    fun hasNext(): Boolean = idx < length

    fun peek(): Char = if (hasNext()) src[idx] else '\u0000'

    fun next(): Char {
        if (!hasNext()) return '\u0000'

        val c = src[idx]
        idx++

        if (c == '\n') {
            line++
            column = 1
        } else if (c == '\r') {
            if (hasNext() && src[idx] == '\n') idx++
            line++
            column = 1
        } else {
            column++
        }

        return c
    }

    fun getSourceLine(targetLine: Int): String? {
        val lineIndex = targetLine - 1
        if (lineIndex !in linesStartIndices.indices) return null
        val startPos = linesStartIndices[lineIndex]
        val endPos = if (lineIndex + 1 < linesStartIndices.size) linesStartIndices[lineIndex + 1] else src.length
        return src.substring(startPos, endPos).trimEnd('\n', '\r')
    }
}

object Tokenizer {
    private val keywordMap: Map<String, TokenType> = mapOf(
        "return" to TokenType.RETURN,
        "if" to TokenType.IF,
        "elif" to TokenType.ELIF,
        "else" to TokenType.ELSE,
        "loop" to TokenType.LOOP,
        "break" to TokenType.BREAK,
        "continue" to TokenType.CONTINUE,
        "init" to TokenType.INIT,
        "this" to TokenType.THIS,
        "import" to TokenType.IMPORT,
        "true" to TokenType.TRUE,
        "false" to TokenType.FALSE,
        "null" to TokenType.NULL,
        "getter" to TokenType.GETTER,
        "setter" to TokenType.SETTER,
        "pub" to TokenType.PUBLIC,
        "pri" to TokenType.PRIVATE,
        "const" to TokenType.CONST
    )

    private fun addToken(tokenList: MutableList<Token>, type: TokenType, info: TokenInfo, literal: String?) {
        tokenList.add(Token(literal, info, type))
    }

    private fun readString(cs: CharStream, info: TokenInfo, quote: Char): String {
        var closed = false
        val content = buildString {
            while (cs.hasNext()) {
                val ch = cs.next()
                if (ch == quote) {
                    closed = true
                    return@buildString
                }
                if (ch == '\\' && cs.hasNext()) {
                    val next = cs.next()
                    append(
                        when (next) {
                            'n' -> '\n'
                            'r' -> '\r'
                            't' -> '\t'
                            '\\', '"', '\'' -> next
                            else -> {
                                append('\\')
                                next
                            }
                        }
                    )
                } else {
                    append(ch)
                }
            }
        }
        if (!closed) {
            ErrorHandler.report(
                info = info,
                reason = "Unterminated String",
                sourceLine = cs.getSourceLine(info.line)
            )
        }
        return content
    }

    private fun readNumber(cs: CharStream, info: TokenInfo, first: Char): Pair<TokenType, String> {
        val sb = StringBuilder()
        sb.append(first)
        while (cs.peek().isDigit()) sb.append(cs.next())
        var isFloat = false
        if (cs.peek() == '.') {
            isFloat = true
            sb.append(cs.next())
            if (!cs.peek().isDigit()) {
                ErrorHandler.report(
                    info = info,
                    reason = "Invalid Float",
                    sourceLine = cs.getSourceLine(info.line)
                )
            } else {
                while (cs.peek().isDigit()) sb.append(cs.next())
            }
        }
        return (if (isFloat) TokenType.FLOAT else TokenType.INT) to sb.toString()
    }

    private fun readIdentifier(cs: CharStream, first: Char): String {
        val id = StringBuilder().append(first)
        while (cs.peek().isLetterOrDigit() || cs.peek() == '_') id.append(cs.next())
        return id.toString()
    }

    private fun handleOperatorOrDelimiter(
        c: Char,
        cs: CharStream,
        info: TokenInfo,
        tokenList: MutableList<Token>
    ): Boolean {
        when (c) {
            '+' -> {
                if (cs.peek() == '=') {
                    cs.next()
                    addToken(tokenList, TokenType.ADD_ASSIGN, info, "+=")
                } else addToken(tokenList, TokenType.ADD, info, "+")
            }
            '-' -> {
                if (cs.peek() == '=') {
                    cs.next()
                    addToken(tokenList, TokenType.MINUS_ASSIGN, info, "-=")
                } else if (cs.peek() == '>') {
                    cs.next()
                    addToken(tokenList, TokenType.ARROW, info, "->")
                } else addToken(tokenList, TokenType.MINUS, info, "-")
            }
            '=' -> {
                if (cs.peek() == '=') {
                    cs.next()
                    addToken(tokenList, TokenType.EQ, info, "==")
                } else addToken(tokenList, TokenType.ASSIGN, info, "=")
            }
            '*' -> {
                if (cs.peek() == '=') {
                    cs.next()
                    addToken(tokenList, TokenType.MULT_ASSIGN, info, "*=")
                } else addToken(tokenList, TokenType.MULT, info, "*")
            }
            '/' -> {
                if (cs.peek() == '=') {
                    cs.next()
                    addToken(tokenList, TokenType.DIV_ASSIGN, info, "/=")
                } else if (cs.peek() == '/') {
                    cs.next()
                    while (cs.hasNext() && cs.peek() != '\n' && cs.peek() != '\r') cs.next()
                } else {
                    addToken(tokenList, TokenType.DIV, info, "/")
                }
            }
            '%' -> {
                if (cs.peek() == '=') {
                    cs.next()
                    addToken(tokenList, TokenType.REM_ASSIGN, info, "%=")
                } else addToken(tokenList, TokenType.REM, info, "%")
            }
            '^' -> addToken(tokenList, TokenType.XOR, info, "^")
            '~' -> addToken(tokenList, TokenType.NOTB, info, "~")
            '&', '|', '>', '<' -> {
                if (cs.peek() == c) {
                    cs.next()
                    val type: TokenType = when (c) {
                        '&' -> TokenType.ANDL
                        '|' -> TokenType.ORL
                        '>' -> TokenType.SHIFTR
                        else -> TokenType.SHIFTL
                    }
                    addToken(tokenList, type, info, "$c$c")
                } else if ((c == '>' || c == '<') && cs.peek() == '=') {
                    cs.next()
                    addToken(tokenList, if (c == '>') TokenType.GE else TokenType.LE, info, "$c=")
                } else {
                    val type: TokenType = when (c) {
                        '&' -> TokenType.ANDB
                        '|' -> TokenType.ORB
                        '>' -> TokenType.GT
                        else -> TokenType.LT
                    }
                    addToken(tokenList, type, info, "$c")
                }
            }
            '!' -> {
                if (cs.peek() == '=') {
                    cs.next()
                    addToken(tokenList, TokenType.NE, info, "!=")
                } else {
                    addToken(tokenList, TokenType.NOTL, info, "!")
                }
            }
            '(' -> addToken(tokenList, TokenType.LPAR, info, "(")
            ')' -> addToken(tokenList, TokenType.RPAR, info, ")")
            '[' -> addToken(tokenList, TokenType.LBRACK, info, "[")
            ']' -> addToken(tokenList, TokenType.RBRACK, info, "]")
            '{' -> addToken(tokenList, TokenType.LBRA, info, "{")
            '}' -> addToken(tokenList, TokenType.RBRA, info, "}")
            ';' -> addToken(tokenList, TokenType.SEM, info, ";")
            ',' -> addToken(tokenList, TokenType.COMMA, info, ",")
            '.' -> addToken(tokenList, TokenType.DOT, info, ".")
            '#' -> addToken(tokenList, TokenType.HASH, info, "#")
            ':' -> addToken(tokenList, TokenType.COLON, info, ":")
            '?' -> addToken(tokenList, TokenType.QUEST, info, "?")
            else -> return false
        }
        return true
    }

    fun tokenize(src: String, fileName: String? = null): MutableList<Token> {
        val tokenList = mutableListOf<Token>()
        val cs = CharStream(src)
        while (cs.hasNext()) {
            val info = TokenInfo(fileName, cs.line, cs.column)
            val c = cs.next()
            when (c) {
                ' ', '\t', '\r', '\n' -> { }
                '/' -> {
                    if (cs.peek() == '/') {
                        cs.next()
                        while (cs.hasNext() && cs.peek() != '\n' && cs.peek() != '\r') cs.next()
                    } else {
                        addToken(tokenList, TokenType.DIV, info, "/")
                    }
                }
                '"', '\'' -> {
                    val content = readString(cs, info, c)
                    addToken(tokenList, TokenType.STRING, info, content)
                }
                else -> {
                    when {
                        c.isDigit() -> {
                            val (type, literal) = readNumber(cs, info, c)
                            addToken(tokenList, type, info, literal)
                        }
                        c.isLetter() || c == '_' -> {
                            val lex = readIdentifier(cs, c)
                            addToken(tokenList, keywordMap[lex] ?: TokenType.ID, info, lex)
                        }
                        handleOperatorOrDelimiter(c, cs, info, tokenList) -> { }
                        else -> ErrorHandler.report(
                            info = info,
                            reason = "Unexpected Char",
                            sourceLine = cs.getSourceLine(info.line)
                        )
                    }
                }
            }
        }
        addToken(tokenList, TokenType.EOF, TokenInfo(fileName, 0, 0), null)
        return tokenList
    }
}

File: src/frontend/lexer/tokenizer.py
from typing import Optional

from .char_stream import CharStream
from .error_handler import ErrorHandler
from .token_types import Token, TokenInfo, TokenType


class Tokenizer:
    _KEYWORD_MAP: dict[str, TokenType] = {
        "return": TokenType.RETURN,
        "if": TokenType.IF,
        "elif": TokenType.ELIF,
        "else": TokenType.ELSE,
        "loop": TokenType.LOOP,
        "break": TokenType.BREAK,
        "continue": TokenType.CONTINUE,
        "init": TokenType.INIT,
        "this": TokenType.THIS,
        "import": TokenType.IMPORT,
        "true": TokenType.TRUE,
        "false": TokenType.FALSE,
        "null": TokenType.NULL,
        "getter": TokenType.GETTER,
        "setter": TokenType.SETTER,
        "pub": TokenType.PUBLIC,
        "pri": TokenType.PRIVATE,
        "const": TokenType.CONST,
        "static": TokenType.STATIC,
        "is": TokenType.IS,
        "try": TokenType.TRY,
        "catch": TokenType.CATCH,
        "finally": TokenType.FINALLY,
        "throw": TokenType.THROW,
    }

    @staticmethod
    def _add_token(
        token_list: list[Token], type: TokenType, info: TokenInfo, literal: Optional[str]
    ) -> None:
        token_list.append(Token(literal, info, type))

    @staticmethod
    def _read_string(cs: CharStream, info: TokenInfo, quote: str) -> str:
        closed = False
        chars = []

        while cs.has_next():
            ch = cs.next()
            if ch == quote:
                closed = True
                break
            if ch == "\\" and cs.has_next():
                nxt = cs.next()
                chars.append(
                    {
                        "n": "\n",
                        "r": "\r",
                        "t": "\t",
                        "\\": "\\",
                        '"': '"',
                        "'": "'",
                    }.get(nxt, f"\\{nxt}")
                )
            else:
                chars.append(ch)

        if not closed:
            ErrorHandler.report(
                info=info, reason="Unterminated String", source_line=cs.get_source_line(info.line)
            )
        return "".join(chars)

    @staticmethod
    def _read_number(cs: CharStream, info: TokenInfo, first: str) -> tuple[TokenType, str]:
        sb = [first]
        while cs.peek().isdigit():
            sb.append(cs.next())

        is_float = False
        if cs.peek() == ".":
            is_float = True
            sb.append(cs.next())
            if not cs.peek().isdigit():
                ErrorHandler.report(
                    info=info, reason="Invalid Float", source_line=cs.get_source_line(info.line)
                )
            else:
                while cs.peek().isdigit():
                    sb.append(cs.next())

        return (TokenType.FLOAT if is_float else TokenType.INT, "".join(sb))

    @staticmethod
    def _read_identifier(cs: CharStream, first: str) -> str:
        sb = [first]
        while cs.peek().isalnum() or cs.peek() == "_":
            sb.append(cs.next())
        return "".join(sb)

    @staticmethod
    def _handle_operator_or_delimiter(
        c: str, cs: CharStream, info: TokenInfo, token_list: list[Token]
    ) -> bool:
        match c:
            case "+":
                if cs.peek() == "=":
                    cs.next()
                    Tokenizer._add_token(token_list, TokenType.ADD_ASSIGN, info, "+=")
                else:
                    Tokenizer._add_token(token_list, TokenType.ADD, info, "+")
            case "-":
                if cs.peek() == "=":
                    cs.next()
                    Tokenizer._add_token(token_list, TokenType.MINUS_ASSIGN, info, "-=")
                elif cs.peek() == ">":
                    cs.next()
                    Tokenizer._add_token(token_list, TokenType.ARROW, info, "->")
                else:
                    Tokenizer._add_token(token_list, TokenType.MINUS, info, "-")
            case "=":
                if cs.peek() == "=":
                    cs.next()
                    Tokenizer._add_token(token_list, TokenType.EQ, info, "==")
                else:
                    Tokenizer._add_token(token_list, TokenType.ASSIGN, info, "=")
            case "*":
                if cs.peek() == "=":
                    cs.next()
                    Tokenizer._add_token(token_list, TokenType.MULT_ASSIGN, info, "*=")
                else:
                    Tokenizer._add_token(token_list, TokenType.MULT, info, "*")
            case "/":
                if cs.peek() == "=":
                    cs.next()
                    Tokenizer._add_token(token_list, TokenType.DIV_ASSIGN, info, "/=")
                elif cs.peek() == "/":
                    cs.next()
                    while cs.has_next() and cs.peek() not in "\n\r":
                        cs.next()
                elif cs.peek() == "*":
                    cs.next()
                    cnt = 1
                    while cs.has_next() and cnt > 0:
                        cur = cs.next()
                        if cur == "*" and cs.peek() == "/":
                            cs.next()
                            cnt -= 1
                        elif cur == "/" and cs.peek() == "*":
                            cs.next()
                            cnt += 1
                else:
                    Tokenizer._add_token(token_list, TokenType.DIV, info, "/")
            case "%":
                if cs.peek() == "=":
                    cs.next()
                    Tokenizer._add_token(token_list, TokenType.REM_ASSIGN, info, "%=")
                else:
                    Tokenizer._add_token(token_list, TokenType.REM, info, "%")
            case "^":
                Tokenizer._add_token(token_list, TokenType.XOR, info, "^")
            case "~":
                Tokenizer._add_token(token_list, TokenType.NOTB, info, "~")
            case "&" | "|" | ">" | "<":
                if cs.peek() == c:
                    cs.next()
                    type_map = {
                        "&": TokenType.ANDL,
                        "|": TokenType.ORL,
                        ">": TokenType.SHIFTR,
                        "<": TokenType.SHIFTL,
                    }
                    Tokenizer._add_token(token_list, type_map[c], info, c + c)
                elif c in "><" and cs.peek() == "=":
                    cs.next()
                    Tokenizer._add_token(
                        token_list, TokenType.GE if c == ">" else TokenType.LE, info, f"{c}="
                    )
                else:
                    type_map = {
                        "&": TokenType.ANDB,
                        "|": TokenType.ORB,
                        ">": TokenType.GT,
                        "<": TokenType.LT,
                    }
                    Tokenizer._add_token(token_list, type_map[c], info, c)
            case "!":
                if cs.peek() == "=":
                    cs.next()
                    Tokenizer._add_token(token_list, TokenType.NE, info, "!=")
                else:
                    Tokenizer._add_token(token_list, TokenType.NOTL, info, "!")
            case "(":
                Tokenizer._add_token(token_list, TokenType.LPAR, info, "(")
            case ")":
                Tokenizer._add_token(token_list, TokenType.RPAR, info, ")")
            case "[":
                Tokenizer._add_token(token_list, TokenType.LBRACK, info, "[")
            case "]":
                Tokenizer._add_token(token_list, TokenType.RBRACK, info, "]")
            case "{":
                Tokenizer._add_token(token_list, TokenType.LBRA, info, "{")
            case "}":
                Tokenizer._add_token(token_list, TokenType.RBRA, info, "}")
            case ";":
                Tokenizer._add_token(token_list, TokenType.SEM, info, ";")
            case ",":
                Tokenizer._add_token(token_list, TokenType.COMMA, info, ",")
            case ".":
                Tokenizer._add_token(token_list, TokenType.DOT, info, ".")
            case "#":
                Tokenizer._add_token(token_list, TokenType.HASH, info, "#")
            case "@":
                Tokenizer._add_token(token_list, TokenType.AT, info, "@")
            case ":":
                Tokenizer._add_token(token_list, TokenType.COLON, info, ":")
            case "?":
                Tokenizer._add_token(token_list, TokenType.QUEST, info, "?")
            case _:
                return False
        return True

    @staticmethod
    def _handle_slash(cs: CharStream, info: TokenInfo, token_list: list[Token]) -> None:
        if cs.peek() == "=":
            cs.next()
            Tokenizer._add_token(token_list, TokenType.DIV_ASSIGN, info, "/=")
        elif cs.peek() == "/":
            cs.next()
            while cs.has_next() and cs.peek() not in "\n\r":
                cs.next()
        elif cs.peek() == "*":
            cs.next()
            cnt = 1
            while cs.has_next() and cnt > 0:
                cur = cs.next()
                if cur == "*" and cs.peek() == "/":
                    cs.next()
                    cnt -= 1
                elif cur == "/" and cs.peek() == "*":
                    cs.next()
                    cnt += 1
        else:
            Tokenizer._add_token(token_list, TokenType.DIV, info, "/")

    @staticmethod
    def tokenize(src: str, file_name: Optional[str] = None) -> list[Token]:
        token_list: list[Token] = []
        cs = CharStream(src)

        while cs.has_next():
            info = TokenInfo(file_name, cs.line, cs.column)
            c = cs.next()

            if c in " \t\r\n":
                continue
            elif c == "/":
                Tokenizer._handle_slash(cs, info, token_list)
            elif c in "\"'":
                content = Tokenizer._read_string(cs, info, c)
                Tokenizer._add_token(token_list, TokenType.STRING, info, content)
            elif c.isdigit():
                ttype, literal = Tokenizer._read_number(cs, info, c)
                Tokenizer._add_token(token_list, ttype, info, literal)
            elif c.isalpha() or c == "_":
                lex = Tokenizer._read_identifier(cs, c)
                ttype = Tokenizer._KEYWORD_MAP.get(lex, TokenType.ID)
                Tokenizer._add_token(token_list, ttype, info, lex)
            elif not Tokenizer._handle_operator_or_delimiter(c, cs, info, token_list):
                ErrorHandler.report(
                    info=info, reason="Unexpected Char", source_line=cs.get_source_line(info.line)
                )

        Tokenizer._add_token(token_list, TokenType.EOF, TokenInfo(file_name, 0, 0), None)
        return token_list
